{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1015857b",
   "metadata": {},
   "source": [
    "# Single cell RNA-seq analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edd3168f",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Import libraries\n",
    "\n",
    "library(Seurat)\n",
    "library(Matrix)\n",
    "library(stringr)\n",
    "library(readr)\n",
    "library(Matrix)\n",
    "library(data.table)\n",
    "library(dplyr)\n",
    "library(biomaRt)\n",
    "library(AUCell)\n",
    "library(ggplot2)\n",
    "library(ropls)\n",
    "library(grid)\n",
    "library(tidyr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e303a51",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Set working directory\n",
    "cwd <- getwd()\n",
    "data <- normalizePath(file.path(cwd, \"../output_files/Single_Cell\"))\n",
    "figures <- normalizePath(file.path(cwd, \"../figures/Single_Cell\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b36b51d",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "# First ggplot2# First we upload the files needed for the creation fo the Seurat object:\n",
    "# - metadata file\n",
    "# - count file\n",
    "# - sample data file\n",
    "# - AUCell score\n",
    "\n",
    "metadata_file <- read.delim(\n",
    "  \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/data/single_cell_rna_seq/GSE162170_rna_cell_metadata.txt\",\n",
    "  header = TRUE,\n",
    "  sep = \"\\t\",\n",
    "  quote = \"\",\n",
    "  comment.char = \"\",\n",
    "  row.names = 1\n",
    ")\n",
    "\n",
    "# we read the counts file with a function that allows a better ulpoading of heavy files\n",
    "counts_file <- fread(\"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/data/single_cell_rna_seq/counts_filtered.tsv\", data.table = FALSE)\n",
    "#  Then we set the first column as the row names and delete the first column\n",
    "rownames(counts_file) <- counts_file[[1]]\n",
    "counts_file[[1]] <- NULL\n",
    "# make the matrix sparse\n",
    "counts_file <- Matrix(as.matrix(counts_file), sparse = TRUE)\n",
    "\n",
    "# Check fi the NEUROD2 is present in our data\n",
    "\"ENSG00000171532\" %in% rownames(counts_file)\n",
    "\n",
    "sample_file <- read.table(\"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/data/single_cell_rna_seq/GSE162170_rna_sample_qc.tsv\")\n",
    "# the same as with the count file\n",
    "colnames(sample_file) <- sample_file[1,]\n",
    "sample_file <- sample_file[-1, ]\n",
    "# the AUCell score for the NEUROD2 for each cell\n",
    "auc_score <- read.csv('/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/data/single_cell_rna_seq/aucell_values_chip.csv')\n",
    "\n",
    "######################################################\n",
    "# Construction of the Seurat object\n",
    "######################################################\n",
    "\n",
    "# load the counts file\n",
    "tiss <- CreateSeuratObject(counts = counts_file)\n",
    "\n",
    "# create a filtered metadata file that contain only the cells that are present after the downsampling\n",
    "metadata_file_filtered <- metadata_file[rownames(metadata_file) %in% colnames(counts_file), ]\n",
    "# check if the cells in both metadata and counts are the same\n",
    "all(rownames(metadata_file_filtered) == colnames(counts_file))\n",
    "\n",
    "# add the metadata to the Seurat object\n",
    "tiss <- AddMetaData(object = tiss, metadata_file_filtered)\n",
    "# subtract the established bashline\n",
    "tiss <- AddMetaData(object = tiss, tiss$Percent.MT, col.name = \"Percent.MT\")\n",
    "# Add the AUCell score to the Seurat object\n",
    "tiss <- AddMetaData(\n",
    "  object = tiss,\n",
    "  metadata = auc_score\n",
    ")\n",
    "\n",
    "######################################################\n",
    "# ENSEMBLE - GENE NAME CORRESPONDANCE\n",
    "######################################################\n",
    "# Let's create a dictionary that will allow us to translate the Ensemble codes to gene names\n",
    "# COnnect to the database to retrieve the name correspondances\n",
    "ensembl <- useMart(\"ensembl\", dataset = \"hsapiens_gene_ensembl\")\n",
    "ensembl <- useEnsembl(\"ensembl\", \n",
    "                      dataset = \"hsapiens_gene_ensembl\", \n",
    "                      mirror = \"us\")  # or \"asia\" / \"www\"\n",
    "# Define which are the genes of which we want to obtain the gene\n",
    "genes <- rownames(counts_file)\n",
    "# Apply the mapping and obtain the respective dataframe\n",
    "gene_map <- getBM(\n",
    "  filters = \"ensembl_gene_id\",\n",
    "  attributes = c(\"ensembl_gene_id\", \"hgnc_symbol\"),\n",
    "  values = genes,\n",
    "  mart = ensembl\n",
    ")\n",
    "# Create a dictionary from the dataframe\n",
    "gene_to_name_dict <- setNames(gene_map$hgnc_symbol, gene_map$ensembl_gene_id)\n",
    "\n",
    "######################################################\n",
    "# Vln plot to check that all goes nice\n",
    "######################################################\n",
    "vln_plot_n_features_counts <- VlnPlot(tiss, features = c(\"nFeature_RNA\", \"nCount_RNA\"), ncol = 2)\n",
    "\n",
    "ggsave(\n",
    "  filename = \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/quality_check_features_counts_vlnplot.tiff\",\n",
    "  plot = vln_plot_n_features_counts,\n",
    "  device = \"tiff\",\n",
    "  dpi = 600,       # publication-quality resolution\n",
    "  width = 8,       # width in inches\n",
    "  height = 6,      # height in inches\n",
    "  units = \"in\",\n",
    "  compression = \"lzw\"  # lossless compression\n",
    ")\n",
    "\n",
    "######################################################\n",
    "# Data transformation\n",
    "######################################################\n",
    "# the combination of this gives scaled and centered data, \n",
    "# which improves the comparison variables with different source biological variability\n",
    "tiss <- NormalizeData(object = tiss, scale.factor = 1e6)\n",
    "tiss <- ScaleData(object = tiss)\n",
    "\n",
    "######################################################\n",
    "# DImensionality reduction stuff\n",
    "######################################################\n",
    "# Run variable feature detection\n",
    "tiss <- FindVariableFeatures(tiss)\n",
    "\n",
    "# Check top variable features\n",
    "top10 <- head(VariableFeatures(tiss), 10)\n",
    "print(top10)\n",
    "\n",
    "# Plot and label\n",
    "plot1 <- VariableFeaturePlot(tiss)\n",
    "plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)\n",
    "plot2\n",
    "# > plot2\n",
    "# Error in grid.Call(C_convert, x, as.integer(whatfrom), as.integer(whatto),  : \n",
    "#                      Viewport has zero dimension(s)\n",
    "# Export to PDF\n",
    "pdf(\"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/variable_features_plot.pdf\", width=8, height=6)\n",
    "plot1\n",
    "dev.off()\n",
    "pdf(\"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/variable_features_plot_with_labels.pdf\", width=8, height=6)\n",
    "plot2\n",
    "dev.off()\n",
    "# To know which are actually the genes\n",
    "top10_symbols <- gene_to_name_dict[top10]\n",
    "# For any missing mappings, keep the original Ensembl ID\n",
    "top10_symbols[is.na(top10_symbols)] <- top10[is.na(top10_symbols)]\n",
    "# ENSG00000078596 ENSG00000184113 ENSG00000276070 ENSG00000275302 ENSG00000276085 ENSG00000118785 ENSG00000277632 \n",
    "# \"ITM2A\"         \"CLDN5\"        \"CCL4L2\"          \"CCL4\"        \"CCL3L3\"          \"SPP1\"          \"CCL3\" \n",
    "# ENSG00000188536 ENSG00000196565 ENSG00000244734 \n",
    "# \"HBA2\"          \"HBG2\"           \"HBB\" \n",
    "\n",
    "# now let's save those genes that are most variable\n",
    "# Use variable genes (found by Seurat)\n",
    "variable_genes <- VariableFeatures(tiss)\n",
    "\n",
    "tiss <- RunPCA(tiss, features = VariableFeatures(object = tiss))\n",
    "ElbowPlot(tiss, ndims = 50)\n",
    "\n",
    "pdf(\"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/elbow_plot_pca.pdf\", width=8, height=6)\n",
    "ElbowPlot(tiss, ndims = 50)\n",
    "dev.off()\n",
    "# so we decided using 20 PCs based on the elbow plot results\n",
    "# Let’s find clusters.\n",
    "tiss <- FindNeighbors(tiss, dims = 1:20)\n",
    "# with this we will use the Louvain algorithm to find clusters\n",
    "tiss <- FindClusters(tiss, resolution = 0.5)\n",
    "# let's group it by Age and Tissue\n",
    "tiss <- RunUMAP(tiss, dims = 1:43)\n",
    "# now we construct the Seurat clusters and save them\n",
    "umap_plot_batch <- DimPlot(tiss, reduction = \"umap\", group.by = 'Batch')\n",
    "umap_plot_age <- DimPlot(tiss, reduction = \"umap\", group.by = 'Age')\n",
    "umap_plot_tissue_id <- DimPlot(tiss, reduction = \"umap\", group.by = 'Tissue.ID')\n",
    "umap_plot_seurat_clusters <- DimPlot(tiss, reduction = \"umap\", group.by = 'seurat_clusters')\n",
    "umap_plot_seurat_clusters_labeled <- DimPlot(tiss, reduction = \"umap\", group.by = 'seurat_clusters', label = TRUE)\n",
    "umap_plot_cell_type <- DimPlot(tiss, reduction = \"umap\", group.by = 'cell_type')\n",
    "umap_plot_cell_type_labeled <- DimPlot(tiss, reduction = \"umap\", group.by = 'cell_type', label = TRUE)\n",
    "\n",
    "ggsave(\n",
    "  filename = \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/umap_plot_batch.tiff\",\n",
    "  plot = umap_plot_batch,\n",
    "  device = \"tiff\",\n",
    "  dpi = 600,       # publication-quality resolution\n",
    "  width = 8,       # width in inches\n",
    "  height = 6,      # height in inches\n",
    "  units = \"in\",\n",
    "  compression = \"lzw\"  # lossless compression\n",
    ")\n",
    "ggsave(\n",
    "  filename = \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/umap_plot_age.tiff\",\n",
    "  plot = umap_plot_age,\n",
    "  device = \"tiff\",\n",
    "  dpi = 600,       # publication-quality resolution\n",
    "  width = 8,       # width in inches\n",
    "  height = 6,      # height in inches\n",
    "  units = \"in\",\n",
    "  compression = \"lzw\"  # lossless compression\n",
    ")\n",
    "ggsave(\n",
    "  filename = \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/umap_plot_tissue_id.tiff\",\n",
    "  plot = umap_plot_tissue_id,\n",
    "  device = \"tiff\",\n",
    "  dpi = 600,       # publication-quality resolution\n",
    "  width = 8,       # width in inches\n",
    "  height = 6,      # height in inches\n",
    "  units = \"in\",\n",
    "  compression = \"lzw\"  # lossless compression\n",
    ")\n",
    "ggsave(\n",
    "  filename = \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/umap_plot_seurat_clusters.tiff\",\n",
    "  plot = umap_plot_seurat_clusters,\n",
    "  device = \"tiff\",\n",
    "  dpi = 600,       # publication-quality resolution\n",
    "  width = 8,       # width in inches\n",
    "  height = 6,      # height in inches\n",
    "  units = \"in\",\n",
    "  compression = \"lzw\"  # lossless compression\n",
    ")\n",
    "ggsave(\n",
    "  filename = \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/umap_plot_seurat_clusters_labeled.tiff\",\n",
    "  plot = umap_plot_seurat_clusters_labeled,\n",
    "  device = \"tiff\",\n",
    "  dpi = 600,       # publication-quality resolution\n",
    "  width = 8,       # width in inches\n",
    "  height = 6,      # height in inches\n",
    "  units = \"in\",\n",
    "  compression = \"lzw\"  # lossless compression\n",
    ")\n",
    "ggsave(\n",
    "  filename = \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/umap_plot_cell_type.tiff\",\n",
    "  plot = umap_plot_cell_type,\n",
    "  device = \"tiff\",\n",
    "  dpi = 600,       # publication-quality resolution\n",
    "  width = 8,       # width in inches\n",
    "  height = 6,      # height in inches\n",
    "  units = \"in\",\n",
    "  compression = \"lzw\"  # lossless compression\n",
    ")\n",
    "ggsave(\n",
    "  filename = \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/umap_plot_cell_type_labeled.tiff\",\n",
    "  plot = umap_plot_cell_type_labeled,\n",
    "  device = \"tiff\",\n",
    "  dpi = 600,       # publication-quality resolution\n",
    "  width = 8,       # width in inches\n",
    "  height = 6,      # height in inches\n",
    "  units = \"in\",\n",
    "  compression = \"lzw\"  # lossless compression\n",
    ")\n",
    "\n",
    "# Now we will procede with feature plots and those things\n",
    "# this is for GAD1, MGP, NEUROD2, TREM2 and the AUCell score\n",
    "f_plot_gad1 <- FeaturePlot(tiss, features = c(\"ENSG00000128683\"))\n",
    "f_plot_mgp <- FeaturePlot(tiss, features = c(\"ENSG00000111341\"))\n",
    "f_plot_neurod2 <- FeaturePlot(tiss, features = c(\"ENSG00000171532\"))\n",
    "f_plot_trem2 <- FeaturePlot(tiss, features = c(\"ENSG00000095970\"))\n",
    "f_plot_aucscore <- FeaturePlot(tiss, features = c(\"geneSet\"))\n",
    "\n",
    "ggsave(\n",
    "  filename = \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/feature_plot_gad1.tiff\",\n",
    "  plot = f_plot_gad1,\n",
    "  device = \"tiff\",\n",
    "  dpi = 600,       # publication-quality resolution\n",
    "  width = 8,       # width in inches\n",
    "  height = 6,      # height in inches\n",
    "  units = \"in\",\n",
    "  compression = \"lzw\"  # lossless compression\n",
    ")\n",
    "ggsave(\n",
    "  filename = \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/feature_plot_mgp.tiff\",\n",
    "  plot = f_plot_mgp,\n",
    "  device = \"tiff\",\n",
    "  dpi = 600,       # publication-quality resolution\n",
    "  width = 8,       # width in inches\n",
    "  height = 6,      # height in inches\n",
    "  units = \"in\",\n",
    "  compression = \"lzw\"  # lossless compression\n",
    ")\n",
    "ggsave(\n",
    "  filename = \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/feature_plot_neurod2.tiff\",\n",
    "  plot = f_plot_neurod2,\n",
    "  device = \"tiff\",\n",
    "  dpi = 600,       # publication-quality resolution\n",
    "  width = 8,       # width in inches\n",
    "  height = 6,      # height in inches\n",
    "  units = \"in\",\n",
    "  compression = \"lzw\"  # lossless compression\n",
    ")\n",
    "ggsave(\n",
    "  filename = \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/feature_plot_trem2.tiff\",\n",
    "  plot = f_plot_trem2,\n",
    "  device = \"tiff\",\n",
    "  dpi = 600,       # publication-quality resolution\n",
    "  width = 8,       # width in inches\n",
    "  height = 6,      # height in inches\n",
    "  units = \"in\",\n",
    "  compression = \"lzw\"  # lossless compression\n",
    ")\n",
    "ggsave(\n",
    "  filename = \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/feature_plot_aucscore.tiff\",\n",
    "  plot = f_plot_aucscore,\n",
    "  device = \"tiff\",\n",
    "  dpi = 600,       # publication-quality resolution\n",
    "  width = 8,       # width in inches\n",
    "  height = 6,      # height in inches\n",
    "  units = \"in\",\n",
    "  compression = \"lzw\"  # lossless compression\n",
    ")\n",
    "\n",
    "######################################################\n",
    "# Find markers genes\n",
    "######################################################\n",
    "\n",
    "Idents(tiss) <- tiss$cell_type\n",
    "celltypes <- unique(Idents(tiss))\n",
    "comparisons <- combn(celltypes, 2, simplify = FALSE)\n",
    "\n",
    "outdir <- \"/home/alberto/Desktop/pairwise_markers\"\n",
    "pairwise_markers <- list()\n",
    "for (pair in comparisons) {\n",
    "  cellA <- pair[1]\n",
    "  cellB <- pair[2]\n",
    "  \n",
    "  message(\"Comparing \", cellA, \" vs \", cellB)\n",
    "  \n",
    "  # sanitize names for filenames\n",
    "  safeA <- str_replace_all(cellA, \"[^A-Za-z0-9_-]\", \"_\")\n",
    "  safeB <- str_replace_all(cellB, \"[^A-Za-z0-9_-]\", \"_\")\n",
    "  \n",
    "  markers <- FindMarkers(\n",
    "    tiss,\n",
    "    ident.1 = cellA,\n",
    "    ident.2 = cellB,\n",
    "    only.pos = FALSE,\n",
    "    min.pct = 0.25,\n",
    "    logfc.threshold = 0.25\n",
    "  )\n",
    "  \n",
    "  # store in memory\n",
    "  pair_name <- paste(cellA, \"_vs_\", cellB, sep = \"\")\n",
    "  pairwise_markers[[pair_name]] <- markers\n",
    "  \n",
    "  # safe output file path\n",
    "  out_path <- file.path(outdir, paste0(safeA, \"_vs_\", safeB, \"_markers.tsv\"))\n",
    "  \n",
    "  write.table(markers, file = out_path, sep = \"\\t\", quote = FALSE, col.names = NA)\n",
    "}\n",
    "\n",
    "################# put together all tables\n",
    "all_markers_df <- bind_rows(\n",
    "  lapply(names(pairwise_markers), function(x) {\n",
    "    df <- pairwise_markers[[x]]\n",
    "    df$comparison <- x\n",
    "    df\n",
    "  })\n",
    ")\n",
    "\n",
    "write.table(all_markers_df, \"/home/alberto/Desktop/pairwise_markers/all_pairwise_markers_cell_type.tsv\", sep=\"\\t\", quote=FALSE, col.names=NA)\n",
    "\n",
    "# now for the age agrupation\n",
    "Idents(tiss) <- tiss$Age   # or use tiss$seurat_clusters\n",
    "celltypes <- unique(Idents(tiss))\n",
    "comparisons <- combn(celltypes, 2, simplify = FALSE)\n",
    "\n",
    "outdir <- \"/home/alberto/Desktop/pairwise_markers\"\n",
    "pairwise_markers <- list()\n",
    "for (pair in comparisons) {\n",
    "  cellA <- pair[1]\n",
    "  cellB <- pair[2]\n",
    "  \n",
    "  message(\"Comparing \", cellA, \" vs \", cellB)\n",
    "  \n",
    "  # sanitize names for filenames\n",
    "  safeA <- str_replace_all(cellA, \"[^A-Za-z0-9_-]\", \"_\")\n",
    "  safeB <- str_replace_all(cellB, \"[^A-Za-z0-9_-]\", \"_\")\n",
    "  \n",
    "  markers <- FindMarkers(\n",
    "    tiss,\n",
    "    ident.1 = cellA,\n",
    "    ident.2 = cellB,\n",
    "    only.pos = FALSE,\n",
    "    min.pct = 0.25,\n",
    "    logfc.threshold = 0.25\n",
    "  )\n",
    "  \n",
    "  # store in memory\n",
    "  pair_name <- paste(cellA, \"_vs_\", cellB, sep = \"\")\n",
    "  pairwise_markers[[pair_name]] <- markers\n",
    "  \n",
    "  # safe output file path\n",
    "  out_path <- file.path(outdir, paste0(safeA, \"_vs_\", safeB, \"_markers.tsv\"))\n",
    "  \n",
    "  write.table(markers, file = out_path, sep = \"\\t\", quote = FALSE, col.names = NA)\n",
    "}\n",
    "\n",
    "################# put together all tables\n",
    "all_markers_df <- bind_rows(\n",
    "  lapply(names(pairwise_markers), function(x) {\n",
    "    df <- pairwise_markers[[x]]\n",
    "    df$comparison <- x\n",
    "    df\n",
    "  })\n",
    ")\n",
    "\n",
    "write.table(all_markers_df, \"/home/alberto/Desktop/pairwise_markers/all_pairwise_markers_age.tsv\", sep=\"\\t\", quote=FALSE, col.names=NA)\n",
    "\n",
    "###################### find unique markers per cell type\n",
    "unique_markers <- list()\n",
    "\n",
    "for (ct in celltypes) {\n",
    "  others <- setdiff(celltypes, ct)\n",
    "  tmp <- list()\n",
    "  \n",
    "  for (other in others) {\n",
    "    res <- FindMarkers(tiss, ident.1 = ct, ident.2 = other, only.pos = TRUE)\n",
    "    tmp[[other]] <- res\n",
    "  }\n",
    "  \n",
    "  # Intersect top markers across all comparisons\n",
    "  shared_genes <- Reduce(intersect, lapply(tmp, rownames))\n",
    "  unique_markers[[ct]] <- shared_genes\n",
    "}\n",
    "\n",
    "unique_markers\n",
    "\n",
    "######################################################\n",
    "# Ridge plots and Violin plots, for enurod2\n",
    "######################################################\n",
    "\n",
    "sc_rdg <- RidgePlot(tiss, features = 'ENSG00000171532', group.by = 'seurat_clusters')\n",
    "ct_rdg <- RidgePlot(tiss, features = 'ENSG00000171532', group.by = 'cell_type')\n",
    "ag_rdg <- RidgePlot(tiss, features = 'ENSG00000171532', group.by = 'Age')\n",
    "tsid_rdg <- RidgePlot(tiss, features = 'ENSG00000171532', group.by = 'Tissue.ID')\n",
    "ba_rdg <- RidgePlot(tiss, features = 'ENSG00000171532', group.by = 'Batch')\n",
    "smid_rdg <- RidgePlot(tiss, features = 'ENSG00000171532', group.by = 'Sample.ID')\n",
    "\n",
    "sc_vln <- VlnPlot(tiss, features = 'ENSG00000171532', group.by = 'seurat_clusters')\n",
    "ct_vln <- VlnPlot(tiss, features = 'ENSG00000171532', group.by = 'cell_type')\n",
    "ag_vln <- VlnPlot(tiss, features = 'ENSG00000171532', group.by = 'Age')\n",
    "tsid_vln <- VlnPlot(tiss, features = 'ENSG00000171532', group.by = 'Tissue.ID')\n",
    "ba_vln <- VlnPlot(tiss, features = 'ENSG00000171532', group.by = 'Batch')\n",
    "smid_vln <- VlnPlot(tiss, features = 'ENSG00000171532', group.by = 'Sample.ID')\n",
    "\n",
    "ggsave(\n",
    "  filename = \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/ridge_neurod2_seuract_clusters.tiff\",\n",
    "  plot = sc_rdg,\n",
    "  device = \"tiff\",\n",
    "  dpi = 600,       # publication-quality resolution\n",
    "  width = 8,       # width in inches\n",
    "  height = 6,      # height in inches\n",
    "  units = \"in\",\n",
    "  compression = \"lzw\"  # lossless compression\n",
    ")\n",
    "\n",
    "ggsave(\n",
    "  filename = \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/ridge_neurod2_cell_type.tiff\",\n",
    "  plot = ct_rdg,\n",
    "  device = \"tiff\",\n",
    "  dpi = 600,       # publication-quality resolution\n",
    "  width = 8,       # width in inches\n",
    "  height = 6,      # height in inches\n",
    "  units = \"in\",\n",
    "  compression = \"lzw\"  # lossless compression\n",
    ")\n",
    "\n",
    "ggsave(\n",
    "  filename = \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/ridge_neurod2_age.tiff\",\n",
    "  plot = ag_rdg,\n",
    "  device = \"tiff\",\n",
    "  dpi = 600,       # publication-quality resolution\n",
    "  width = 8,       # width in inches\n",
    "  height = 6,      # height in inches\n",
    "  units = \"in\",\n",
    "  compression = \"lzw\"  # lossless compression\n",
    ")\n",
    "\n",
    "ggsave(\n",
    "  filename = \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/ridge_neurod2_tissue_id.tiff\",\n",
    "  plot = tsid_rdg,\n",
    "  device = \"tiff\",\n",
    "  dpi = 600,       # publication-quality resolution\n",
    "  width = 8,       # width in inches\n",
    "  height = 6,      # height in inches\n",
    "  units = \"in\",\n",
    "  compression = \"lzw\"  # lossless compression\n",
    ")\n",
    "\n",
    "ggsave(\n",
    "  filename = \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/ridge_neurod2_batch.tiff\",\n",
    "  plot = ba_rdg,\n",
    "  device = \"tiff\",\n",
    "  dpi = 600,       # publication-quality resolution\n",
    "  width = 8,       # width in inches\n",
    "  height = 6,      # height in inches\n",
    "  units = \"in\",\n",
    "  compression = \"lzw\"  # lossless compression\n",
    ")\n",
    "\n",
    "ggsave(\n",
    "  filename = \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/ridge_neurod2_sample_id.tiff\",\n",
    "  plot = smid_rdg,\n",
    "  device = \"tiff\",\n",
    "  dpi = 600,       # publication-quality resolution\n",
    "  width = 8,       # width in inches\n",
    "  height = 6,      # height in inches\n",
    "  units = \"in\",\n",
    "  compression = \"lzw\"  # lossless compression\n",
    ")\n",
    "\n",
    "# ###############################################\n",
    "ggsave(\n",
    "  filename = \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/vln_neurod2_seuract_clusters.tiff\",\n",
    "  plot = sc_vln,\n",
    "  device = \"tiff\",\n",
    "  dpi = 600,       # publication-quality resolution\n",
    "  width = 8,       # width in inches\n",
    "  height = 6,      # height in inches\n",
    "  units = \"in\",\n",
    "  compression = \"lzw\"  # lossless compression\n",
    ")\n",
    "\n",
    "ggsave(\n",
    "  filename = \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/vln_neurod2_cell_type.tiff\",\n",
    "  plot = ct_vln,\n",
    "  device = \"tiff\",\n",
    "  dpi = 600,       # publication-quality resolution\n",
    "  width = 8,       # width in inches\n",
    "  height = 6,      # height in inches\n",
    "  units = \"in\",\n",
    "  compression = \"lzw\"  # lossless compression\n",
    ")\n",
    "\n",
    "ggsave(\n",
    "  filename = \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/vln_neurod2_age.tiff\",\n",
    "  plot = ag_vln,\n",
    "  device = \"tiff\",\n",
    "  dpi = 600,       # publication-quality resolution\n",
    "  width = 8,       # width in inches\n",
    "  height = 6,      # height in inches\n",
    "  units = \"in\",\n",
    "  compression = \"lzw\"  # lossless compression\n",
    ")\n",
    "\n",
    "ggsave(\n",
    "  filename = \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/vln_neurod2_tissue_id.tiff\",\n",
    "  plot = tsid_vln,\n",
    "  device = \"tiff\",\n",
    "  dpi = 600,       # publication-quality resolution\n",
    "  width = 8,       # width in inches\n",
    "  height = 6,      # height in inches\n",
    "  units = \"in\",\n",
    "  compression = \"lzw\"  # lossless compression\n",
    ")\n",
    "\n",
    "ggsave(\n",
    "  filename = \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/vln_neurod2_batch.tiff\",\n",
    "  plot = ba_vln,\n",
    "  device = \"tiff\",\n",
    "  dpi = 600,       # publication-quality resolution\n",
    "  width = 8,       # width in inches\n",
    "  height = 6,      # height in inches\n",
    "  units = \"in\",\n",
    "  compression = \"lzw\"  # lossless compression\n",
    ")\n",
    "\n",
    "ggsave(\n",
    "  filename = \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/vln_neurod2_sample_id.tiff\",\n",
    "  plot = smid_vln,\n",
    "  device = \"tiff\",\n",
    "  dpi = 600,       # publication-quality resolution\n",
    "  width = 8,       # width in inches\n",
    "  height = 6,      # height in inches\n",
    "  units = \"in\",\n",
    "  compression = \"lzw\"  # lossless compression\n",
    ")\n",
    "\n",
    "######################################################\n",
    "# Percentage of cells expressing an specific gene\n",
    "######################################################\n",
    "\n",
    "# gene of interest (NEUROD2)\n",
    "gene <- \"ENSG00000171532\"\n",
    "\n",
    "# make sure metadata has Sample.ID and Tissue.ID\n",
    "meta <- tiss@meta.data\n",
    "\n",
    "# 1) percent expressing, mean, median per tissue\n",
    "expr_mat <- as.matrix(GetAssayData(tiss, assay = \"RNA\", slot = \"counts\")) # raw counts\n",
    "if(!gene %in% rownames(expr_mat)) stop(\"Gene not found in counts matrix\")\n",
    "\n",
    "df_summary_cell_type <- meta %>%\n",
    "  mutate(cell = rownames(meta),\n",
    "         expr = as.numeric(expr_mat[gene, cell])) %>%\n",
    "  # group_by(cell_type) %>%\n",
    "  group_by(cell_type) %>%\n",
    "  # group_by(CTxA) %>%\n",
    "  summarize(\n",
    "    n_cells = n(),\n",
    "    n_samples = n_distinct(Sample.ID),\n",
    "    pct_expr = 100 * mean(expr > 0),\n",
    "    mean_expr = mean(expr),\n",
    "    mean_expr_in_expr_cells = ifelse(sum(expr>0)>0, mean(expr[expr>0]), NA),\n",
    "    median_expr = median(expr)\n",
    "  ) %>%\n",
    "  arrange(desc(mean_expr))\n",
    "\n",
    "df_summary_age <- meta %>%\n",
    "  mutate(cell = rownames(meta),\n",
    "         expr = as.numeric(expr_mat[gene, cell])) %>%\n",
    "  # group_by(cell_type) %>%\n",
    "  group_by(Age) %>%\n",
    "  # group_by(CTxA) %>%\n",
    "  summarize(\n",
    "    n_cells = n(),\n",
    "    n_samples = n_distinct(Sample.ID),\n",
    "    pct_expr = 100 * mean(expr > 0),\n",
    "    mean_expr = mean(expr),\n",
    "    mean_expr_in_expr_cells = ifelse(sum(expr>0)>0, mean(expr[expr>0]), NA),\n",
    "    median_expr = median(expr)\n",
    "  ) %>%\n",
    "  arrange(desc(mean_expr))\n",
    "\n",
    "# optional pretty table\n",
    "write.table(df_summary_cell_type, \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/neurod2_cell_type_summary.tsv\", sep=\"\\t\", row.names=FALSE, quote=FALSE)\n",
    "\n",
    "write.table(df_summary_age, \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/neurod2_age_summary.tsv\", sep=\"\\t\", row.names=FALSE, quote=FALSE)\n",
    "\n",
    "vln_cell_type_neurod2 <- VlnPlot(tiss, features = gene, group.by = \"cell_type\", pt.size = 0.5) + NoLegend()\n",
    "vln_age_neurod2 <- VlnPlot(tiss, features = gene, group.by = \"Age\", pt.size = 0.5) + NoLegend()\n",
    "\n",
    "cell_type_percentage_expressing_neurod2 <- ggplot(df_summary_cell_type, aes(x = reorder(cell_type, -pct_expr), y = pct_expr)) +\n",
    "  geom_col() + coord_flip() + ylab(\"% cells expressing\") + xlab(\"Tissue\") +\n",
    "  theme_minimal()\n",
    "age_percentage_expressing_neurod2 <- ggplot(df_summary_age, aes(x = reorder(Age, -pct_expr), y = pct_expr)) +\n",
    "  geom_col() + coord_flip() + ylab(\"% cells expressing\") + xlab(\"Tissue\") +\n",
    "  theme_minimal()\n",
    "\n",
    "ggsave(\n",
    "  filename = \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/vln_neurod2_per_cell_type.tiff\",\n",
    "  plot = vln_cell_type_neurod2,\n",
    "  device = \"tiff\",\n",
    "  dpi = 600,       # publication-quality resolution\n",
    "  width = 8,       # width in inches\n",
    "  height = 6,      # height in inches\n",
    "  units = \"in\",\n",
    "  compression = \"lzw\"  # lossless compression\n",
    ")\n",
    "\n",
    "ggsave(\n",
    "  filename = \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/ratio_neurod2_per_cell_type.tiff\",\n",
    "  plot = cell_type_percentage_expressing_neurod2,\n",
    "  device = \"tiff\",\n",
    "  dpi = 600,       # publication-quality resolution\n",
    "  width = 8,       # width in inches\n",
    "  height = 6,      # height in inches\n",
    "  units = \"in\",\n",
    "  compression = \"lzw\"  # lossless compression\n",
    ")\n",
    "\n",
    "ggsave(\n",
    "  filename = \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/vln_neurod2_per_age.tiff\",\n",
    "  plot = vln_age_neurod2,\n",
    "  device = \"tiff\",\n",
    "  dpi = 600,       # publication-quality resolution\n",
    "  width = 8,       # width in inches\n",
    "  height = 6,      # height in inches\n",
    "  units = \"in\",\n",
    "  compression = \"lzw\"  # lossless compression\n",
    ")\n",
    "\n",
    "ggsave(\n",
    "  filename = \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/ratio_neurod2_per_age.tiff\",\n",
    "  plot = age_percentage_expressing_neurod2,\n",
    "  device = \"tiff\",\n",
    "  dpi = 600,       # publication-quality resolution\n",
    "  width = 8,       # width in inches\n",
    "  height = 6,      # height in inches\n",
    "  units = \"in\",\n",
    "  compression = \"lzw\"  # lossless compression\n",
    ")\n",
    "\n",
    "######################################################\n",
    "# Sassy plots for NEUROD2 relative expression in each cell and each cell type\n",
    "######################################################\n",
    "\n",
    "# Extrae las coordenadas UMAP + metadatos + expresión de NeuroD2\n",
    "gene <- \"ENSG00000171532\"\n",
    "umap_df <- as.data.frame(Embeddings(tiss, \"umap\")) %>%\n",
    "  mutate(\n",
    "    cell = rownames(.),\n",
    "    cell_type = tiss$cell_type,\n",
    "    neurod2 = FetchData(tiss, vars = gene)[,1]\n",
    "  )\n",
    "\n",
    "# UMAP coloreado por expresión, con elipses de cell_type\n",
    "plot_umap_neurod2_cooler <- ggplot(umap_df, aes(umap_1, umap_2)) +\n",
    "  geom_point(aes(color = neurod2), size = 0.4, alpha = 0.8) +\n",
    "  # stat_ellipse(aes(group = cell_type), color = \"black\", lwd = 0.4, linetype = \"dashed\") +\n",
    "  scale_color_gradientn(colors = c(\"grey90\", \"royalblue\", \"darkred\")) +\n",
    "  theme_minimal(base_size = 12) +\n",
    "  theme(\n",
    "    legend.position = \"right\",\n",
    "    panel.grid = element_blank()\n",
    "  ) +\n",
    "  labs(\n",
    "    color = \"NEUROD2\\nexpression\",\n",
    "    title = \"UMAP colored by NEUROD2 expression\\nEllipses show cell-type clusters\"\n",
    "  )\n",
    "\n",
    "ggsave(\n",
    "  filename = \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/umap_neurod2_cooler.tiff\",\n",
    "  plot = plot_umap_neurod2_cooler,\n",
    "  device = \"tiff\",\n",
    "  dpi = 600,       # publication-quality resolution\n",
    "  width = 8,       # width in inches\n",
    "  height = 6,      # height in inches\n",
    "  units = \"in\",\n",
    "  compression = \"lzw\"  # lossless compression\n",
    ")\n",
    "\n",
    "centroids <- umap_df %>%\n",
    "  group_by(cell_type) %>%\n",
    "  summarise(\n",
    "    umap_1 = mean(umap_1),\n",
    "    umap_2 = mean(umap_2)\n",
    "  )\n",
    "\n",
    "# Crear el plot\n",
    "plot_umap_neurod2_cooler_cell_type_labeled <- ggplot(umap_df, aes(umap_1, umap_2)) +\n",
    "  geom_point(aes(color = neurod2), size = 0.4, alpha = 0.8) +\n",
    "  stat_ellipse(aes(group = cell_type), color = \"black\", lwd = 0.4, linetype = \"dashed\") +\n",
    "  geom_text(\n",
    "    data = centroids,\n",
    "    aes(label = cell_type),\n",
    "    color = \"black\",\n",
    "    size = 3,\n",
    "    fontface = \"bold\"\n",
    "  ) +\n",
    "  scale_color_gradientn(colors = c(\"grey90\", \"royalblue\", \"darkred\")) +\n",
    "  theme_minimal(base_size = 12) +\n",
    "  theme(\n",
    "    legend.position = \"right\",\n",
    "    panel.grid = element_blank()\n",
    "  ) +\n",
    "  labs(\n",
    "    color = \"NEUROD2\\nexpression\",\n",
    "    title = \"UMAP colored by NEUROD2 expression\\nEllipses show cell-type clusters\"\n",
    "  )\n",
    "\n",
    "ggsave(\n",
    "  filename = \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/umap_neurod2_cooler_cell_type_labeled.tiff\",\n",
    "  plot = plot_umap_neurod2_cooler_cell_type_labeled,\n",
    "  device = \"tiff\",\n",
    "  dpi = 600,       # publication-quality resolution\n",
    "  width = 8,       # width in inches\n",
    "  height = 6,      # height in inches\n",
    "  units = \"in\",\n",
    "  compression = \"lzw\"  # lossless compression\n",
    ")\n",
    "\n",
    "######################################################\n",
    "# OPLS pipeline\n",
    "######################################################\n",
    "\n",
    "counts_file <- fread(\"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/data/single_cell_rna_seq/counts_filtered.tsv\", data.table = FALSE)\n",
    "#  Then we set the first column as the row names and delete the first column\n",
    "rownames(counts_file) <- counts_file[[1]]\n",
    "counts_file[[1]] <- NULL\n",
    "\n",
    "# Normalized data\n",
    "gene <- \"ENSG00000171532\"\n",
    "\n",
    "if (!gene %in% rownames(counts_file)) {\n",
    "  stop(\"❌ NEUROD2 (ENSG00000171532) not found in the data file!\")\n",
    "}\n",
    "\n",
    "# define the variable that we want to predict (in this case, NEUROD2 expression)\n",
    "y_cont <- as.numeric(counts_file[gene, ])\n",
    "\n",
    "# Define X = all other genes (cells × genes)\n",
    "X <- t(counts_file[setdiff(rownames(counts_file), gene), ])  # transpose to cells × genes\n",
    "\n",
    "# Run OPLS\n",
    "# define the agrupations we want to apply in the permutation\n",
    "cv_segments <- min(5, nrow(X) - 1)\n",
    "\n",
    "opls_res <- opls(\n",
    "  X,\n",
    "  y_cont,\n",
    "  predI = 1,\n",
    "  orthoI = 10,\n",
    "  crossvalI = cv_segments,\n",
    "  scaleC = c(\"none\", \"center\", \"pareto\", \"standard\")[4],\n",
    "  permI = 0\n",
    ")\n",
    "\n",
    "# OPLS\n",
    "# 7918 samples x 2000 variables and 1 response\n",
    "# standard scaling of predictors and response(s)\n",
    "# R2X(cum) R2Y(cum) Q2(cum) RMSEE pre ort\n",
    "# Total    0.539    0.624    0.56  1.84   1  10\n",
    "\n",
    "# # Get variable importance scores\n",
    "# vip_scores <- getVipVn(opls_res)\n",
    "# vip_df <- data.frame(\n",
    "#   gene = names(vip_scores),\n",
    "#   VIP = vip_scores\n",
    "# )\n",
    "# vip_df <- vip_df[order(vip_df$VIP, decreasing = TRUE), ]\n",
    "# \n",
    "# head(vip_df, 20)\n",
    "\n",
    "pred_VIP <- opls_res@vipVn\n",
    "ortho_VIP <- opls_res@orthoVipVn\n",
    "\n",
    "keys <- names(pred_VIP)\n",
    "pred_values <- as.numeric(pred_VIP)\n",
    "ortho_values <- as.numeric(ortho_VIP)\n",
    "\n",
    "\n",
    "vip_opls_df <- data.frame(\n",
    "  variable = keys,\n",
    "  pred = pred_values, \n",
    "  ortho = ortho_values\n",
    ")\n",
    "\n",
    "vip_filtered_df <- vip_opls_df[vip_opls_df$pred > 2,]\n",
    "\n",
    "# pred_VIP[pred_VIP > 1]\n",
    "# cat(names(pred_VIP[pred_VIP > 1]), sep = \", \")\n",
    "# \n",
    "# ortho_VIP[ortho_VIP > 1]\n",
    "# cat(names(ortho_VIP[ortho_VIP > 1]), sep = \", \")\n",
    "# \n",
    "# opls_res@coefficientMN\n",
    "# opls_res@loadingMN\n",
    "\n",
    "p1 <- opls_res@loadingMN[, 1]       # p[1] loading (covariance)\n",
    "pcorr <- opls_res@coefficientMN[, 1]        # p(corr)[1] (correlation)\n",
    "names <- rownames(opls_res@loadingMN)  # variable names\n",
    "\n",
    "# Build the dataframe\n",
    "s_data <- data.frame(\n",
    "  Variable = names,\n",
    "  pcorr = pcorr\n",
    ")\n",
    "\n",
    "s_data_filtered <- s_data[s_data$Variable %in% vip_filtered_df$variable, ]\n",
    "\n",
    "p1 <- opls_res@scoreMN[, 1]    \n",
    "o1 <- opls_res@orthoScoreMN[, 1]\n",
    "\n",
    "pred_ortho_df <- data.frame(\n",
    "  variable = y_cont,\n",
    "  p1 = p1,\n",
    "  o1 = o1\n",
    ")\n",
    "\n",
    "write.csv(pred_ortho_df, paste0(\"/home/alberto/Desktop/pred_ortho.csv\"), row.names = FALSE)\n",
    "\n",
    "write.csv(vip_filtered_df, paste0(\"/home/alberto/Desktop/variates_pred_ortho_sc.csv\"), row.names = FALSE)\n",
    "\n",
    "write.csv(s_data_filtered, paste0(\"/home/alberto/Desktop/corr_to_splot_sc.csv\"), row.names = FALSE)\n",
    "#################################\n",
    "pred_ortho_df <- read.csv(paste0(\"/home/alberto/Desktop/pred_ortho.csv\"))\n",
    "\n",
    "vip_filtered_df <- read.csv(paste0(\"/home/alberto/Desktop/variates_pred_ortho_sc.csv\"))\n",
    "\n",
    "s_data_filtered <- read.csv(paste0(\"/home/alberto/Desktop/corr_to_splot_sc.csv\"))\n",
    "\n",
    "# ---------------------------\n",
    "\n",
    "scores_plot_opls <- ggplot(pred_ortho_df, aes(x = p1, y = o1, color = variable)) +\n",
    "  geom_point(size = 3) +\n",
    "  # geom_text(vjust = -0.5, size = 3) +\n",
    "  stat_ellipse(\n",
    "    type = \"norm\", \n",
    "    level = 0.95,\n",
    "    color = \"#262626\",       # Border color\n",
    "    fill = \"#EBEBEB\",         # Interior fill color\n",
    "    alpha = 0.1,             # Transparency of fill\n",
    "    linewidth = 1,\n",
    "    geom = \"polygon\"         # Important: needed for filled ellipses\n",
    "  )+\n",
    "  scale_color_gradient(  low = \"#1C9C89\",     # azul verdoso\n",
    "                         # mid = \"#D3D3D3\",     # gris claro (puedes ajustar)\n",
    "                         high = \"#643173\",    # morado\n",
    "                         # name = \"IL-10 (pg/mL)\"\n",
    "                         # name = \"CRP (ng/mL)\"\n",
    "  ) +  # Puedes personalizar los colores\n",
    "  theme_minimal() +\n",
    "  labs(title = \"\", x = \"p1\", y = \"o1\") +\n",
    "  coord_cartesian(xlim = c(-7.5, 7.5), ylim = c(-10, 10))+\n",
    "  theme(\n",
    "    plot.title = element_text(face = \"bold\", size = 14, hjust = 0.5),\n",
    "    axis.title = element_text(size = 12),\n",
    "    panel.grid.major = element_line(color = \"grey80\")\n",
    "  )+\n",
    "  # geom_hline(yintercept = 0, color = \"#999999\", linewidth = 0.6, linetype = \"dashed\") +\n",
    "  # geom_vline(xintercept = 0, color = \"#999999\", linewidth = 0.6, linetype = \"dashed\")\n",
    "  theme(\n",
    "    axis.line.x = element_line(color = \"#7A7A7A\", linewidth = 0.75),\n",
    "    axis.line.y = element_line(color = \"#7A7A7A\", linewidth = 0.75),\n",
    "    axis.title.x = element_text(face = \"bold.italic\", size = 20, hjust = 0.5,color='#262626'),  # Bold X-axis title\n",
    "    axis.text.x = element_text(size = 20, hjust = 0.5,color='#262626'),  # Bold X-axis title\n",
    "    axis.text.y = element_text(size = 20, hjust = 0.5,color='#262626'),  # Bold X-axis title\n",
    "    axis.title.y = element_text(face = \"bold.italic\", size = 20, hjust = 0.5,color='#262626'),  # Bold Y-axis title\n",
    "    plot.title = element_text(face = \"bold\", size = 20, hjust = 0.5, color = '#262626'),\n",
    "    panel.grid.major = element_line(color = \"grey80\"),\n",
    "    legend.title = element_text(size = 16, face = \"bold\", color = \"#262626\"),\n",
    "    legend.text  = element_text(size = 14, face = \"bold\", color = \"#262626\"),\n",
    "    legend.key.size = unit(1.2, \"cm\")\n",
    "    \n",
    "  ) +\n",
    "  annotation_custom(\n",
    "    grob = textGrob(\n",
    "      # IL-10\n",
    "      label = paste(\"R2 = 0.624\\nQ2 = 0.560\"),\n",
    "      # CRP\n",
    "      # label = paste(\"R2 = 0.635\\nQ2 = 0.575\"),\n",
    "      # this line of beneath allows to control de position of this label with the p-value and the rho value\n",
    "      x = 0.981, y = 0.90, hjust = 1,\n",
    "      gp = gpar(col = \"#262626\", fontsize = 13, fontface = \"bold\", fill = 'white')\n",
    "    )\n",
    "  )\n",
    "\n",
    "ggsave(\n",
    "  filename = \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/scores_plot_opls.tiff\",\n",
    "  plot = scores_plot_opls,\n",
    "  device = \"tiff\",\n",
    "  dpi = 600,       # publication-quality resolution\n",
    "  width = 8,       # width in inches\n",
    "  height = 6,      # height in inches\n",
    "  units = \"in\",\n",
    "  compression = \"lzw\"  # lossless compression\n",
    ")\n",
    "\n",
    "\n",
    "\n",
    "#########################\n",
    "\n",
    "# Transformar a formato largo\n",
    "vip_long <- vip_filtered_df %>%\n",
    "  pivot_longer(cols = c(pred, ortho),\n",
    "               names_to = \"Component\",\n",
    "               values_to = \"VIP\")\n",
    "\n",
    "# Add a new column with gene symbols\n",
    "vip_long$gene_symbol <- gene_to_name_dict[vip_long$variable]\n",
    "# Remove names from the vector\n",
    "vip_long$gene_symbol <- unname(vip_long$gene_symbol)\n",
    "\n",
    "# Replace missing gene symbols with original Ensembl ID\n",
    "vip_long$gene_symbol[is.na(vip_long$gene_symbol)] <- vip_long$variable\n",
    "\n",
    "pred_order <- vip_long %>%\n",
    "  filter(Component == \"pred\") %>%\n",
    "  arrange(desc(VIP)) %>%\n",
    "  pull(gene_symbol)\n",
    "\n",
    "vip_long$gene_symbol <- factor(vip_long$gene_symbol, levels = pred_order)\n",
    "# Replace missing ones with the original ID\n",
    "# vip_long$gene_symbol[is.na(vip_long$gene_symbol)] <- vip_long$variable\n",
    "\n",
    "# Hacer el gráfico de barras\n",
    "ggplot(vip_long, aes(x = VIP, y = gene_symbol, fill = Component)) +\n",
    "  geom_col(position = \"dodge\") +\n",
    "  labs(x = \"VIP score\", y = \"Gene\", fill = \"Component\") +\n",
    "  theme_minimal(base_size = 12) +\n",
    "  theme(axis.text.y = element_text(size = 6))\n",
    "\n",
    "ggsave(\n",
    "  filename = \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/scores_plot_opls.tiff\",\n",
    "  plot = scores_plot_opls,\n",
    "  device = \"tiff\",\n",
    "  dpi = 600,       # publication-quality resolution\n",
    "  width = 8,       # width in inches\n",
    "  height = 6,      # height in inches\n",
    "  units = \"in\",\n",
    "  compression = \"lzw\"  # lossless compression\n",
    ")\n",
    "\n",
    "###########################################################\n",
    "###########################################################\n",
    "###########################################################\n",
    "\n",
    "# Add a color column based on correlation sign\n",
    "s_data_filtered <- s_data_filtered %>%\n",
    "  mutate(Color = ifelse(pcorr < 0, \"Negative\", \"Positive\"))\n",
    "\n",
    "s_data_filtered$gene_symbol <- gene_to_name_dict[s_data_filtered$Variable]\n",
    "# Remove names from the vector\n",
    "s_data_filtered$gene_symbol <- unname(s_data_filtered$gene_symbol)\n",
    "\n",
    "group_colors <- c(\"Negative\" = \"#5AA9E6\", \"Positive\" = \"orange\")\n",
    "\n",
    "# Plot: Lollipop-style VIP plot\n",
    "vip_scores_sorted_colored <- ggplot(s_data_filtered, aes(x = VIP, y = reorder(gene_symbol, VIP), color = Color)) +\n",
    "  geom_segment(aes(x = 0.95, xend = VIP, y = gene_symbol, yend = gene_symbol), size = 1) +\n",
    "  geom_point(size = 4) +\n",
    "  \n",
    "  scale_color_manual(values = group_colors, name = \"Correlation\") +\n",
    "  \n",
    "  scale_x_continuous(expand = expansion(mult = c(0, 0.02)), limits = c(0.95, NA)) +\n",
    "  \n",
    "  labs(\n",
    "    title = \"VIP Scores by Gene\",\n",
    "    x = \"VIP Score\",\n",
    "    y = \"\"\n",
    "  ) +\n",
    "  theme_minimal(base_size = 14) +\n",
    "  theme(\n",
    "    legend.position = c(0.85, 0.1),\n",
    "    legend.title = element_text(size = 12, face = \"bold\", color = \"#262626\"), \n",
    "    legend.text  = element_text(size = 12, face = \"bold\", color = \"#262626\"),\n",
    "    legend.key = element_blank(),  \n",
    "    legend.key.height = unit(0.3, \"cm\"),\n",
    "    legend.key.width  = unit(0.8, \"cm\"),\n",
    "    \n",
    "    axis.title.x = element_text(face = \"bold\", size = 16, color = '#262626', margin = margin(t=15)),\n",
    "    axis.text.y  = element_text(size = 12, color = '#262626', margin = margin(r=5)),\n",
    "    axis.text.x  = element_text(size = 14, color = '#262626', margin = margin(t=7)),\n",
    "    \n",
    "    panel.grid.major = element_blank(),\n",
    "    panel.grid.minor = element_blank(),\n",
    "    axis.line.x = element_line(color = \"#262626\", size = 1),\n",
    "    axis.line.y = element_line(color = \"#262626\", size = 1)\n",
    "  ) +\n",
    "  guides(\n",
    "    color = guide_legend(override.aes = list(shape = 16, size = 5))\n",
    "  )\n",
    "\n",
    "ggsave(\n",
    "  filename = \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/vip_scores_sorted_colored_opls.tiff\",\n",
    "  plot = vip_scores_sorted_colored,\n",
    "  device = \"tiff\",\n",
    "  dpi = 600,       # publication-quality resolution\n",
    "  width = 13,       # width in inches\n",
    "  height = 13,      # height in inches\n",
    "  units = \"in\",\n",
    "  compression = \"lzw\"  # lossless compression\n",
    ")\n",
    "\n",
    "######################################################\n",
    "# enrichment and pathway analysis\n",
    "######################################################\n",
    "\n",
    "# Install and load gprofiler2 if needed\n",
    "# install.packages(\"gprofiler2\")\n",
    "library(gprofiler2)\n",
    "\n",
    "# Suppose your filtered gene list is in s_data_filtered\n",
    "gene_list <- s_data_filtered$Variable\n",
    "\n",
    "# Perform enrichment analysis\n",
    "gost_res <- gost(\n",
    "  query = gene_list,\n",
    "  organism = \"hsapiens\",   # use appropriate organism code, e.g., \"mmusculus\" for mouse\n",
    "  source = c(\"GO\")  # you can include other sources if you want\n",
    ")\n",
    "\n",
    "# View results\n",
    "head(gost_res$result)\n",
    "\n",
    "# Optional: visualize top enriched GO terms\n",
    "gostplot(gost_res, capped = TRUE, interactive = TRUE)\n",
    "\n",
    "# # Extract the results table\n",
    "# res <- gost_res$result\n",
    "\n",
    "write.csv(as.matrix(gost_res$result), paste0(\"/home/alberto/Desktop/enrichment_sc_rna.csv\"), row.names = FALSE)\n",
    "\n",
    "res_sc_enrichment <- read.csv(\"/home/alberto/Desktop/enrichment_sc_rna.csv\")\n",
    "\n",
    "############################\n",
    "\n",
    "df <- res_sc_enrichment\n",
    "\n",
    "# Filter to keep only significant terms (optional) and p_value\n",
    "df_plot <- df %>% \n",
    "  filter(significant == TRUE) %>%\n",
    "  mutate(log_p = -log10(p_value))\n",
    "\n",
    "# df_plot <- df_plot[df_plot$log_p > 16,]\n",
    "\n",
    "# df_neuro <- df_plot %>%\n",
    "#   filter(\n",
    "#     grepl(\"neuro|nervous|differentiation\", term_name, ignore.case = TRUE)\n",
    "#   )\n",
    "\n",
    "# Plot\n",
    "# Define color palette for GO sources (you can tweak as you wish)\n",
    "source_colors <- c(\n",
    "  \"GO:BP\" = \"#5EB1BF\",  # teal\n",
    "  \"GO:MF\" = \"#F7CE3B\",  # warm yellow\n",
    "  \"GO:CC\" = \"#643173\"   # violet\n",
    ")\n",
    "\n",
    "# Plot with enhanced visual style\n",
    "\n",
    "enrichment_go_processes <- ggplot(df_plot, aes(\n",
    "  x = log_p,\n",
    "  y = reorder(term_name, log_p),\n",
    "  color = source\n",
    ")) +\n",
    "  # ---- thin lines like lollipops ----\n",
    "geom_segment(aes(x = 0, xend = log_p, y = term_name, yend = term_name),\n",
    "             linewidth = 1.2, color = \"grey50\") +\n",
    "  \n",
    "  # ---- dot at the end (optional for clarity) ----\n",
    "geom_point(size = 3, stroke = 0.8) +\n",
    "  \n",
    "  scale_color_manual(values = source_colors) +\n",
    "  \n",
    "  scale_x_continuous(\n",
    "    expand = expansion(mult = c(0, 0.02))  # make sure lines touch y-axis\n",
    "  ) +\n",
    "  \n",
    "  labs(\n",
    "    x = expression(-log[10](\"p-value\")),\n",
    "    y = \"\",\n",
    "    color = \"GO Source\"\n",
    "  ) +\n",
    "  theme_minimal(base_size = 14) +\n",
    "  theme(\n",
    "    axis.line.x = element_line(color = \"#262626\", linewidth = 1),\n",
    "    axis.line.y = element_line(color = \"#262626\", linewidth = 1),\n",
    "    \n",
    "    axis.title.x = element_text(color = \"#645D5E\", face = \"bold\", size = 20, margin = margin(t=15)),\n",
    "    axis.text.x  = element_text(size = 16, color = \"#262626\", margin = margin(t=7)),\n",
    "    axis.text.y  = element_markdown(size = 12, color = \"#262626\", lineheight = 1.4, face = 'bold'),\n",
    "    \n",
    "    panel.grid.major   = element_blank(),\n",
    "    panel.grid.minor   = element_blank(),\n",
    "    panel.border       = element_blank(),\n",
    "    legend.position    = \"right\",\n",
    "    legend.title       = element_text(face = \"bold\", color = \"#262626\", size = 16),\n",
    "    legend.text        = element_text(size = 14, color = \"#262626\"),\n",
    "    plot.margin        = margin(10, 20, 10, 5),\n",
    "    plot.background    = element_rect(fill = \"white\", color = NA),\n",
    "    panel.background   = element_rect(fill = \"white\", color = NA)\n",
    "  ) +\n",
    "  guides(\n",
    "    color = guide_legend(\n",
    "      override.aes = list(size = 5)\n",
    "    )\n",
    "  )\n",
    "\n",
    "ggsave(\n",
    "  filename = \"/media/alberto/External HD1/UPF/master_academic_years/first_master_academic_year/f_m_a_y_first_term/f_m_a_y_f_t_subjects/PGB_principles_of_genomes_bioinformatics/PGB_project/figures/enrichment_go_bubble_plot.tiff\",\n",
    "  plot = enrichment_go_processes,\n",
    "  device = \"tiff\",\n",
    "  dpi = 600,       # publication-quality resolution\n",
    "  width = 13,       # width in inches\n",
    "  height = 13,      # height in inches\n",
    "  units = \"in\",\n",
    "  compression = \"lzw\"  # lossless compression\n",
    ")\n",
    "\n",
    "#########################################################################\n",
    "\n",
    "# This is fro the biochemical pathway data, with KEGG as target database\n",
    "s_data_filtered <- read.csv(\"/home/alberto/Desktop/corr_to_splot_sc.csv\")\n",
    "\n",
    "# Suppose your filtered gene list is in s_data_filtered\n",
    "gene_list <- s_data_filtered$Variable\n",
    "\n",
    "# Perform enrichment analysis\n",
    "kegg_res <- gost(\n",
    "  query = gene_list,\n",
    "  organism = \"hsapiens\",   # use appropriate organism code, e.g., \"mmusculus\" for mouse\n",
    "  source = c(\"KEGG\")  # you can include other sources if you want\n",
    ")\n",
    "\n",
    "write.csv(as.matrix(kegg_res$result), paste0(\"/home/alberto/Desktop/kegg_sc_rna.csv\"), row.names = FALSE)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
